### ISSUE TITLE
[SPIR-V] Struct deep copy workaround fails

### ISSUE THREAD

**Issue by Themaister** [KhronosGroup/glslang#506]
This shader triggers complicated cases with struct deep copy workaround recently implemented.

```
#version 310 es
layout(local_size_x = 1) in;

struct S0
{
    vec2 a[1];
    float b;
};

struct S1
{
    vec3 a;
    float b;
};

struct S2
{
    vec3 a[1];
    float b;
};

struct S3
{
    vec2 a;
    float b;
};

struct Content
{
    S0 m0s[1];
    S1 m1s[1];
    S2 m2s[1];
    S0 m0;
    S1 m1;
    S2 m2;
    S3 m3;
    float m4;
};

layout(binding = 1, std430) buffer SSBO1
{
    Content content;
    Content content1[2];
    Content content2;

    layout(column_major) mat2 m0;
    layout(column_major) mat2 m1;
    layout(column_major) mat2x3 m2[4];
    layout(column_major) mat3x2 m3;
    layout(row_major) mat2 m4;
    layout(row_major) mat2 m5[9];
    layout(row_major) mat2x3 m6[4][2];
    layout(row_major) mat3x2 m7;
    float array[];
} ssbo_430;

layout(binding = 0, std140) buffer SSBO0
{
    Content content;
    Content content1[2];
    Content content2;
    float array[];
} ssbo_140;

void main()
{
    ssbo_430.content = ssbo_140.content;
}
```

SPIRV-Cross generates this code, and glslang then complains about out-of-bounds access:
```
void main()
{
    ssbo_430.content.m0s[0] = ssbo_140.content.m0s[0];
    ssbo_430.content.m0s[1] = ssbo_140.content.m0s[1]; <-- OOB
    ssbo_430.content.m1s[0] = ssbo_140.content.m1s[0];
    ssbo_430.content.m1s[1] = ssbo_140.content.m1s[1];
    ssbo_430.content.m2s[0] = ssbo_140.content.m2s[0];
    ssbo_430.content.m2s[1] = ssbo_140.content.m2s[1];
    ssbo_430.content.m0.a = ssbo_140.content.m0.a;
    ssbo_430.content.m0.b = ssbo_140.content.m0.b;
    ssbo_430.content.m1.a = ssbo_140.content.m1.a;
    ssbo_430.content.m1.b = ssbo_140.content.m1.b;
    ssbo_430.content.m2.a = ssbo_140.content.m2.a;
    ssbo_430.content.m2.b = ssbo_140.content.m2.b;
    ssbo_430.content.m3.a = ssbo_140.content.m3.a;
    ssbo_430.content.m3.b = ssbo_140.content.m3.b;
    ssbo_430.content.m4 = ssbo_140.content.m4;
}
```

SPIR-V snippet
```
       %main = OpFunction %void None %3
          %5 = OpLabel
         %54 = OpAccessChain %_ptr_Uniform_Content_0 %ssbo_140 %37
         %55 = OpLoad %Content_0 %54
         %57 = OpAccessChain %_ptr_Uniform_Content %ssbo_430 %37
         %58 = OpCompositeExtract %_arr_S0_0_9 %55 0
         %60 = OpAccessChain %_ptr_Uniform__arr_S0_9 %57 %37
         %61 = OpCompositeExtract %S0_0 %58 0 <-- Extracts first element
         %63 = OpAccessChain %_ptr_Uniform_S0 %60 %37
               OpStore %63 %61
         %64 = OpCompositeExtract %S0_0 %58 1 <-- Second
         %66 = OpAccessChain %_ptr_Uniform_S0 %60 %65
               OpStore %66 %64
```

**Comment by johnkslang** [comment]
All the issues with struct copy also apply to arrays.  It's really an aggregate issue, but the array solution was missing, getting things a bit mixed up.  Fixed now by b3e24e4.

**Comment by Themaister** [comment]
Confirmed working.
